### PR TITLE
Basic Semantic Segmentation Training Script

### PR DESCRIPTION
**Pull request by DavidLandup0**
# What does this PR do?
As we're working towards introducing more semantic segmentation architectures:

- DeepLabV3 refactor #1119
- DeepLabV3Plus #1128
- FCN #1176

We should have a generalized semantic segmentation training script, similar to the `basic_training.py` for image classification. The general outline for the DeepLabV3 script is great, and we can plug and play models into it, as long as we can customize some aspects of it such as the batch size, initial learning rate, etc.

This script is a generalized form of the existing DeepLabV3 training script meant for any given semantic segmentation architecture on Pascal VOC, in the style of @ianstenbit's `basic_training.py` script. It's been refactored to simplify methods and calls, use KPLs, and use Ian's addition for EMA and weight decay. (currently commented out because they're from an unsupported Keras release afaik)

IMO, having a central script to test these will also force us to ensure that all semantic segmentation models have the same input/output, and that they follow a standardized form that users can change by simply changing the model's name. It's better to introduce this earlier rather than later. Also, this will prevent us having multiple training scripts that are largely the same.

### TODO
A few edge cases left to tweak and introduce, but the majority of the script is there.

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [] Did you write any new necessary tests?
- [] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?

@tanzhenyu @ianstenbit @LukeWood

### CODE REVIEW

File: examples/training/semantic_segmentation/pascal_voc/basic_segmentation_training.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+def preprocess_image(img, cls_seg):
+    img = tf.keras.layers.Resizing(512, 512, interpolation="nearest")(img)
+    inputs = {"images": img, "segmentation_masks": cls_seg}
+    inputs = keras_cv.layers.preprocessing.RandomFlip("horizontal")(inputs)
+    return inputs["images"], inputs["segmentation_masks"]
+
+
+def proc_train_fn(examples):
+    image = examples.pop("image")
```
Comment (DavidLandup0) [inline examples/training/semantic_segmentation/pascal_voc/basic_segmentation_training.py:110]: Maybe flip and then resize instead? 
Since we're getting a dict, popping, turning into a dict again to do `RandomFlip` and then returning separate outputs again.

The former might be more computationally expensive though, in which case this approach is better?

File: examples/training/semantic_segmentation/pascal_voc/basic_segmentation_training.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+    image, cls_seg = preprocess_image(image, cls_seg, flip=True)
+
+    sample_weight = tf.equal(cls_seg, 255)
+    zeros = tf.zeros_like(cls_seg)
+    cls_seg = tf.where(sample_weight, zeros, cls_seg)
+    return image, cls_seg
+
+
+def proc_eval_fn(examples):
```
Comment (DavidLandup0) [inline examples/training/semantic_segmentation/pascal_voc/basic_segmentation_training.py:121]: Since the only difference between `proc_train_fn` and `proc_eval_fn` is whether we're flipping the images as well or not (i.e. do augs or not) - can we collapse the two into a single func and pass an 'augment=boolean' flag?

Anticipating other augmentations that KCV now supports with segmentation masks.

**Comment by DavidLandup0** [comment]
Btw, the preprocessing call works as expected:

**Comment by DavidLandup0** [comment]
Loving the KPLs that support segmentation maps! Added a few placeholder augmentations to test. We should standardize which augmentations are used and to which degree:

image

File: examples/training/semantic_segmentation/pascal_voc/basic_segmentation_training.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        inputs = preprocessing.RandomGaussianBlur(kernel_size=3, factor=0.3)(inputs)
+        inputs = preprocessing.RandomRotation(factor=0.1, segmentation_classes=21)(
+            inputs
+        )
+
+    return inputs["images"], inputs["segmentation_masks"]
+
+
+def proc_train_fn(examples):
+    image = examples.pop("image")
```
Comment (DavidLandup0) [inline examples/training/semantic_segmentation/pascal_voc/basic_segmentation_training.py:120]: Maybe we should just pass `examples` here and pop in the preprocessing function instead?

**Comment by DavidLandup0** [comment]
Doing a 30-epoch run for validation with:

```
!python keras-cv/examples/training/semantic_segmentation/pascal_voc/basic_segmentation_training.py --model_name=DeepLabV3 --backbone=resnet50_v2 --use_ema=False --weights_path='/weights' --tensorboard_path='/tensorboard'
```

**Review by tanzhenyu — CHANGES_REQUESTED** [review]
Thanks for the PR!
Can you help us build incremental changes, i.e.
1) git mv `examples/training/semantic_segmentation/pascal_voc/deeplab.py` `examples/training/semantic_semgnetation/pascal_voc/basic_training.py`
2) only add `model_name` flags, without adding any other KPLs / flags
3) start adding KPLs in next PR, but always run the model end to end to verify before and after effect

**Comment by DavidLandup0** [comment]
We can do step-by-step, but through a series of PRs then, we'd have DeepLab-tuned parameters being used for a general script. Is that okay? Since PRs can take some time to run/verify/merge, won't we create a bottleneck for testing segmentation models with a series of merges?

If this ends up running >= original script, would that be enough validation, to hopefully avoid a bottleneck?

**Comment by tanzhenyu** [comment]
> We can do step-by-step, but through a series of PRs then, we'd have DeepLab-tuned parameters being used for a general script. Is that okay? Since PRs can take some time to run/verify/merge, won't we create a bottleneck for testing segmentation models with a series of merges?
> 
> If this ends up running >= original script, would that be enough validation, to hopefully avoid a bottleneck? I'll go with your gut feeling.

which part of it is specific to deeplab?

**Comment by DavidLandup0** [comment]
> which part of it is specific to deeplab?

The epochs to train for, LR, image sizes, augmentations, batch size, backbone and LR decay are dependent on or tuned to work with DeepLab, and likely won't translate well to other architectures. Similar to how each model trained with `basic_training.py` used *some* different params for training, for any other model here, we'll probably want to set different params.

If we set them all as customizable, and re-run the same configuration DeepLab was trained on (set using flags instead of hardcoded), would that serve as validation? 🤔

**Comment by tanzhenyu** [comment]
> > which part of it is specific to deeplab?
> 
> The epochs to train for, LR, image sizes, augmentations, batch size, backbone and LR decay are dependent on or tuned to work with DeepLab, and likely won't translate well to other architectures. Similar to how each model trained with `basic_training.py` used _some_ different params for training, for any other model here, we'll probably want to set different params.
> 
> If we set them all as customizable, and re-run the same configuration DeepLab was trained on (set using flags instead of hardcoded), would that serve as validation? 🤔

I think image size / batch size / weight decay are usually using common hyperparameters, they might be able to share. For epochs, we could just set for 100 epochs with early stopping.
In the long run, we want to keep training script as generic as possible, and all hyperparameters per model somewhere else, so the training will look like:
`.../basic_training.py`
run_deeplab_train.sh -- call basic_training with its own hyperparameter
run_fcn_train.sh -- call basic_training with its own hyperparameter
etc.

**Comment by tanzhenyu** [comment]
Let's get this PR going by simply scoping it to add `model_name` in the script, and git mv from deeplab_v3.py to basic_training.py, so we could more easily track the diff.
If you wouldn't be able to run git mv to show the diff, feel free to simply open another PR and close this one

**Comment by DavidLandup0** [comment]
I'll likely draft this one and create a new [1/X step] PR with just model_name

**Comment by DavidLandup0** [comment]
Saved changes on a new branch and will repurpose this one for step-by-step

**Comment by DavidLandup0** [comment]
Added model_name flag and kwargs @tanzhenyu

File: examples/training/semantic_segmentation/pascal_voc/basic_training.py [hunk 4]
```diff
@@ -136,15 +148,16 @@ def proc_eval_fn(examples):
...
+    backbone = models.ResNet50V2(
         include_rescaling=True,
         # This argument gives a 2% mIoU increase
         stackwise_dilations=[1, 1, 1, 2],
         input_shape=(512, 512, 3),
         include_top=False,
         weights="imagenet",
     )
-    model = DeepLabV3(classes=21, backbone=backbone)
+    model = segmentation.__dict__[FLAGS.model_name]
```
Comment (tanzhenyu) [inline examples/training/semantic_segmentation/pascal_voc/basic_training.py:159]: this can be retrieved from keras_cv.models.__dict__, no?

**Review by tanzhenyu — CHANGES_REQUESTED** [review]

File: examples/training/semantic_segmentation/pascal_voc/basic_training.py (hunk 4 above)
Comment (DavidLandup0) [reply examples/training/semantic_segmentation/pascal_voc/basic_training.py:159]: Good catch! They weren't originally in `models`, but have been moved since

Comment (DavidLandup0) [reply examples/training/semantic_segmentation/pascal_voc/basic_training.py]: Updated

**Review by tanzhenyu — APPROVED** [review]

**Comment by DavidLandup0** [comment]
Opening next PR as soon as this is merged. Here's the idea for updates:

- Support TPU with mixed precision (currently only works on GPUs)
- Support backbones via flags
- Change preprocessing functions to use KCV layers and simplify the calls 
- Allow changing loss functions
- EMA and weight decay (2.11)
 
Thoughts?

**Comment by tanzhenyu** [comment]
> Opening next PR as soon as this is merged. Here's the idea for updates order:
> 
> * Epochs flag
> * Support TPU with mixed precision (currently only works on GPUs)
> * Support backbones via flags [requires running a test]
> * Change preprocessing functions to use KCV layers and simplify the calls [requires running a test]
> * Allow changing loss functions
> * EMA and weight decay (2.11)
> 
> Thoughts?

Out of all of them, supporting TPU is the most prioritized one. I have test on TPU, it works extremely slow. It might be due to dataset problem or upsampling problem, which I haven't had the chance to figure it out

**Comment by tanzhenyu** [comment]
No need to run gcp test since this only touches training script

**Comment by DavidLandup0** [comment]
> Out of all of them, supporting TPU is the most prioritized one. I have test on TPU, it works extremely slow. It might be due to dataset problem or upsampling problem, which I haven't had the chance to figure it out

Could it be the missing bfloat16? Pushing that change now so we can test whether that helps

**Comment by tanzhenyu** [comment]
> > Out of all of them, supporting TPU is the most prioritized one. I have test on TPU, it works extremely slow. It might be due to dataset problem or upsampling problem, which I haven't had the chance to figure it out
> 
> Could it be the missing bfloat16? Pushing that change now so we can test whether that helps

That's not the reason. TPU will use bfloat by default
